File: apps/agents/src/open-canvas/nodes/rewriteCodeArtifactTheme.ts
import { v4 as uuidv4 } from "uuid";
import { LangGraphRunnableConfig } from "@langchain/langgraph";
import {
  extractThinkingAndResponseTokens,
  isThinkingModel,
} from "@opencanvas/shared/utils/thinking";
import {
  isArtifactCodeContent,
  getArtifactContent,
} from "@opencanvas/shared/utils/artifacts";
import { ArtifactCodeV3, ArtifactV3 } from "@opencanvas/shared/types";
import { getModelConfig, getModelFromConfig } from "../../utils.js";
import {
  ADD_COMMENTS_TO_CODE_ARTIFACT_PROMPT,
  ADD_LOGS_TO_CODE_ARTIFACT_PROMPT,
  FIX_BUGS_CODE_ARTIFACT_PROMPT,
  PORT_LANGUAGE_CODE_ARTIFACT_PROMPT,
} from "../prompts.js";
import {
  OpenCanvasGraphAnnotation,
  OpenCanvasGraphReturnType,
} from "../state.js";
import { AIMessage } from "@langchain/core/messages";

export const rewriteCodeArtifactTheme = async (
  state: typeof OpenCanvasGraphAnnotation.State,
  config: LangGraphRunnableConfig
): Promise<OpenCanvasGraphReturnType> => {
  const { modelName } = getModelConfig(config);
  const smallModel = await getModelFromConfig(config);

  const currentArtifactContent = state.artifact
    ? getArtifactContent(state.artifact)
    : undefined;
  if (!currentArtifactContent) {
    throw new Error("No artifact found");
  }
  if (!isArtifactCodeContent(currentArtifactContent)) {
    throw new Error("Current artifact content is not code");
  }

  let formattedPrompt = "";
  if (state.addComments) {
    formattedPrompt = ADD_COMMENTS_TO_CODE_ARTIFACT_PROMPT;
  } else if (state.portLanguage) {
    let newLanguage = "";
    switch (state.portLanguage) {
      case "typescript":
        newLanguage = "TypeScript";
        break;
      case "javascript":
        newLanguage = "JavaScript";
        break;
      case "cpp":
        newLanguage = "C++";
        break;
      case "java":
        newLanguage = "Java";
        break;
      case "php":
        newLanguage = "PHP";
        break;
      case "python":
        newLanguage = "Python";
        break;
      case "html":
        newLanguage = "HTML";
        break;
      case "sql":
        newLanguage = "SQL";
        break;
    }
    formattedPrompt = PORT_LANGUAGE_CODE_ARTIFACT_PROMPT.replace(
      "{newLanguage}",
      newLanguage
    );
  } else if (state.addLogs) {
    formattedPrompt = ADD_LOGS_TO_CODE_ARTIFACT_PROMPT;
  } else if (state.fixBugs) {
    formattedPrompt = FIX_BUGS_CODE_ARTIFACT_PROMPT;
  } else {
    throw new Error("No theme selected");
  }

  // Insert the code into the artifact placeholder in the prompt
  formattedPrompt = formattedPrompt.replace(
    "{artifactContent}",
    currentArtifactContent.code
  );

  const newArtifactValues = await smallModel.invoke([
    { role: "user", content: formattedPrompt },
  ]);

  let thinkingMessage: AIMessage | undefined;
  let artifactContentText = newArtifactValues.content as string;

  if (isThinkingModel(modelName)) {
    const { thinking, response } =
      extractThinkingAndResponseTokens(artifactContentText);
    thinkingMessage = new AIMessage({
      id: `thinking-${uuidv4()}`,
      content: thinking,
    });
    artifactContentText = response;
  }

  const newArtifactContent: ArtifactCodeV3 = {
    index: state.artifact.contents.length + 1,
    type: "code",
    title: currentArtifactContent.title,
    // Ensure the new artifact's language is updated, if necessary
    language: state.portLanguage || currentArtifactContent.language,
    code: artifactContentText,
  };

  const newArtifact: ArtifactV3 = {
    ...state.artifact,
    currentIndex: state.artifact.contents.length + 1,
    contents: [...state.artifact.contents, newArtifactContent],
  };

  return {
    artifact: newArtifact,
    messages: [...(thinkingMessage ? [thinkingMessage] : [])],
    _messages: [...(thinkingMessage ? [thinkingMessage] : [])],
  };
};

File: agents-py/agents/src/open_canvas/nodes/rewrite_code_artifact_theme.py
import uuid
from typing import Dict, Any
from langchain_core.messages import AIMessage
from langchain_core.runnables import RunnableConfig
from agents.src.open_canvas.state import OpenCanvasGraphState
from shared.src.types import ArtifactV3, ArtifactCodeV3
from agents.src.utils import get_model_config, get_model_from_config
from agents.src.open_canvas.prompts import (
    ADD_COMMENTS_TO_CODE_ARTIFACT_PROMPT,
    ADD_LOGS_TO_CODE_ARTIFACT_PROMPT,
    FIX_BUGS_CODE_ARTIFACT_PROMPT,
    PORT_LANGUAGE_CODE_ARTIFACT_PROMPT
)

async def rewrite_code_artifact_theme(
    state: OpenCanvasGraphState,
    config: RunnableConfig
) -> Dict[str, Any]:
    model_config = get_model_config(config)
    small_model = await get_model_from_config(config)
    
    current_artifact_content = None
    if state.artifact and state.artifact.contents:
        current_artifact_content = state.artifact.contents[-1]
        if not isinstance(current_artifact_content, ArtifactCodeV3):
            raise ValueError("Current artifact content is not code")

    if not current_artifact_content:
        raise ValueError("No artifact found")

    language_map = {
        "typescript": "TypeScript",
        "javascript": "JavaScript",
        "cpp": "C++",
        "java": "Java",
        "php": "PHP",
        "python": "Python",
        "html": "HTML",
        "sql": "SQL"
    }

    formatted_prompt = ""
    if state.add_comments:
        formatted_prompt = ADD_COMMENTS_TO_CODE_ARTIFACT_PROMPT
    elif state.port_language:
        new_lang = language_map.get(state.port_language, state.port_language)
        formatted_prompt = PORT_LANGUAGE_CODE_ARTIFACT_PROMPT.format(
            new_language=new_lang
        )
    elif state.add_logs:
        formatted_prompt = ADD_LOGS_TO_CODE_ARTIFACT_PROMPT
    elif state.fix_bugs:
        formatted_prompt = FIX_BUGS_CODE_ARTIFACT_PROMPT
    else:
        raise ValueError("No theme selected")

    formatted_prompt = formatted_prompt.format(
        artifact_content=current_artifact_content.code
    )

    response = await small_model.invoke([{"role": "user", "content": formatted_prompt}])
    
    new_content = response.content
    thinking_message = None
    
    if "thinking" in model_config.get("model_name", "").lower():
        # Simple thinking/content separation
        thinking_part, _, content_part = new_content.partition("\n\n")
        thinking_message = AIMessage(
            id=f"thinking-{uuid.uuid4()}",
            content=thinking_part
        )
        new_content = content_part

    new_artifact_content = ArtifactCodeV3(
        index=len(state.artifact.contents) + 1,
        type="code",
        title=current_artifact_content.title,
        language=state.port_language or current_artifact_content.language,
        code=new_content
    )

    new_artifact = ArtifactV3(
        current_index=len(state.artifact.contents) + 1,
        contents=[*state.artifact.contents, new_artifact_content]
    )

    return {
        "artifact": new_artifact,
        "messages": [thinking_message] if thinking_message else [],
        "_messages": [thinking_message] if thinking_message else []
    } 